### PR TITLE
Tooltip/popover: Allow viewport option to be a function

### PR DESCRIPTION
**Pull request by Intrepidd**
It might be really useful to be able to pass a function to the viewport option in some cases, just like that :

``` javascript
$('body').tooltip({
  selector: '[data-toggle="tooltip"]',
  viewport: function($element) {
    var viewport = $element.closest('.tooltip-viewport');
    return (viewport.length > 0 ? viewport : $('body'));
  }
});
```

I implemented that in this PR.

I'm open to suggestions, I'm not sure if it's a good idea to accept either a string, an object, or a function. (Did not update the doc yet because of that)

Regarding tests, I noticed that there are no tests asserting that using functions instead of string work (placement for instance), so I copied an existing viewport test, you tell me if that's enough.

Thanks

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 337ce52cc5b03a1d06df963cd3823bf5568a94c3
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55826002

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/tooltip.js [hunk 1]
```diff
@@ -50,7 +50,12 @@
     this.type      = type
     this.$element  = $(element)
     this.options   = this.getOptions(options)
-    this.$viewport = this.options.viewport && $(this.options.viewport.selector || this.options.viewport)
+
+    if (typeof this.options.viewport == 'function') {
+      this.$viewport = this.options.viewport.call(this, this.$element)
+    }
+
+    this.$viewport = this.$viewport || (this.options.viewport && $(this.options.viewport.selector || this.options.viewport))
```
Comment (hnrch02) [inline js/tooltip.js]:
```js
this.$viewport = this.options.viewport && $($.isFunction(this.options.viewport) ? this.options.viewport.call(this, this.$element) : (this.options.viewport.selector || this.options.viewport))
```

would be shorter and would also support DOM nodes being returned.

Comment (Intrepidd) [inline js/tooltip.js]: Thanks, updated

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: cc57ce063dcd51ffbe6ce3503f0522e033033680
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/56105931

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: fb79df3a9b2c3b37398409d8d0cc7011abc31d24
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/56106133

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 043f180c1d053420cae8c892d21475f613c70116
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/56106194

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
* The tooltip+popover documentation needs a corresponding update
* A (nearly-identical) unit test needs to be added to the Popover plugin's test suite

File: js/tests/unit/tooltip.js [hunk 2]
```diff
@@ -719,6 +719,36 @@ $(function () {
...
+        + '</style>'
+    var $styles = $(styles).appendTo('head')
+
+    var $container = $('<div class="container-viewport"/>').appendTo(document.body)
+    var $target = $('<a href="#" rel="tooltip" title="tip" style="top: 50px; left: 350px;"/>')
+      .appendTo($container)
+      .bootstrapTooltip({
+        placement: 'bottom',
+        viewport: function ($element) {
+          return ($element.closest('.container-viewport'))
```
Comment (cvrebert) [inline js/tests/unit/tooltip.js]: Perhaps I'm paranoid, but adding an assertion that `$element[0]` is the correct element might be nice.

**Comment by Intrepidd** [comment]
Thank you!

**Pull request closed without merging**